File: be-scraper-fastapi/app/models/team.py
from sqlmodel import SQLModel, Field
from sqlalchemy import Column
from sqlalchemy.types import JSON
from sqlalchemy.dialects.postgresql import ARRAY, INTEGER
from typing import Optional
from datetime import datetime, timezone
import uuid

class Team(SQLModel, table=True):
    id: Optional[uuid.UUID] = Field(default_factory=uuid.uuid4, primary_key=True)
    name: str
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    deleted_at: Optional[datetime] = None
    description: str
    stage: Optional[str] = None # University, High School, Middle School, etc.
    institution_name: Optional[str] = None
    member_count: Optional[int] = None
    tap_members: Optional[list[uuid.UUID]] = Field(default=None, sa_column=Column(JSON))
    members_list: Optional[list[str]] = Field(default=None, sa_column=Column(JSON))
    leader: Optional[uuid.UUID] = None
    competition_id: uuid.UUID
    years: list[str] = Field(default_factory=list, sa_column=Column(ARRAY(JSON)))
    status: Optional[str] = None # "finalist", "derece", "etc"
    rank: Optional[int] = None
    relation: Optional[str] = None # central, related, friend, None
    intro_file_path: Optional[str] = None
    team_link: Optional[str] = None

File: be-main-hono/src/types/team.ts
export interface Team {
  id: string;
  name: string;
  created_at: string;
  updated_at: string;
  deleted_at: string | null;
  description: string | null;
  stage: string | null;
  institution_name: string | null;
  member_count: number | null;
  tap_members: string[]; // UUIDs
  members_list: string[]; // names
  leader: string | null; // UUID
  competition_id: number;
  years: number[];
  status: string | null;
  rank: number | null;
  relation: string | null;
  intro_file_path: string | null;
  team_link: string | null;
}

export interface CreateTeamPayload {
  id?: string; // if not provided, generated
  name: string;
  description?: string | null;
  stage?: string | null;
  institution_name?: string | null;
  member_count?: number | null;
  tap_members?: string[];
  members_list?: string[];
  leader?: string | null;
  competition_id: number;
  years?: number[];
  status?: string | null;
  rank?: number | null;
  relation?: string | null;
  intro_file_path?: string | null;
  team_link?: string | null;
}

export type UpdateTeamPayload = Partial<CreateTeamPayload>;

export const mapTeamRow = (row: any): Team => {
  const parseJSON = (v: any): any[] => {
    try {
      return v ? JSON.parse(v) : [];
    } catch {
      return [];
    }
  };
  return {
    id: row.id,
    name: row.name,
    created_at: row.created_at,
    updated_at: row.updated_at,
    deleted_at: row.deleted_at ?? null,
    description: row.description ?? null,
    stage: row.stage ?? null,
    institution_name: row.institution_name ?? null,
    member_count: row.member_count != null ? Number(row.member_count) : null,
    tap_members: parseJSON(row.tap_members),
    members_list: parseJSON(row.members_list),
    leader: row.leader ?? null,
    competition_id: Number(row.competition_id),
    years: parseJSON(row.years),
    status: row.status ?? null,
    rank: row.rank != null ? Number(row.rank) : null,
    relation: row.relation ?? null,
    intro_file_path: row.intro_file_path ?? null,
    team_link: row.team_link ?? null,
  };
};
